dashboard/app/helpers/aichat_comprehend_helper.rb

```diff
@@ -8,19 +11,71 @@ def self.create_comprehend_client
 
   # Moderate given text for inappropriate/toxic content using AWS Comprehend client.
   def self.get_toxicity(text, locale)
-    comprehend_response = create_comprehend_client.detect_toxic_content(
-      {
-        text_segments: [{text: text}],
-        language_code: locale,
-      }
-    )
-    categories = comprehend_response.result_list[0].labels
+    text_segment_lists = get_text_segment_lists(text)
+    max_toxicity = 0
+    max_category = nil
+    text_segment_lists.each do |list|
+      comprehend_response = create_comprehend_client.detect_toxic_content(
+        {
+          text_segments: list.map {|segment| {text: segment}},
+          language_code: locale,
+        }
+      )
+      max_toxicity = [max_toxicity, comprehend_response.result_list.map(&:toxicity).max].max
+      list_max_category = comprehend_response.result_list.map(&:labels).flatten.max_by(&:score)
+      max_category = max_category.nil? ? list_max_category : [max_category, list_max_category].max_by(&:score)
+    end
+
     {
       text: text,
-      toxicity: comprehend_response.result_list[0].toxicity,
-      max_category: categories.max_by(&:score),
+      toxicity: max_toxicity,
+      max_category: max_category
     }
   end
+
+  # Returns a collection of text segment lists, each with a list of individual text segments to check for toxicity.
+  # Comprehend limits the amount of bytes per segment, and the amount of overall bytes per list.
+  def self.get_text_segment_lists(text)
+    words = []
+    # Split the text by words. If a single word exceeds our segment size limit, split up the word.
+    text.split.each do |word|
+      if word.bytesize >= MAX_SEGMENT_SIZE_BYTES
+        word_segments = [""]
+        word.each_char do |char|
+          if word_segments.last.bytesize + char.bytesize < MAX_SEGMENT_SIZE_BYTES
+            word_segments.last << char
+          else
+            word_segments << char
+          end
+        end
+        words.concat(word_segments)
+      else
+        words << word
+      end
+    end
+
+    text_segment_lists = [[""]]
+    words.each do |word|
```

Comment:
We could do this in a way that doesn't require processing the text twice—once to split it into words and again to chunk it into segments—and instead does it in a single pass, yeah? But it's probably not worth doing 🙃

Reply:
Its complicated, I think most ways of expressing this would only be more confusing. You could iterate over each byte as an outer loop and keep track of whether you were inside a character or not, but.... I'm not sure that would actually b efaster in ruby.

Reply:
I was thinking that too but I just don't know Ruby well enough 😆 to Seth's point though, seems like it might be simpler to do it as two separate passes anyway?